# Context

Comment spam has grown strongly in recent years on the channels of many major YouTube creators.

Several of them expressed their frustration with malicious comments impersonating them in order to scam their viewers.

This therefore becomes an important problem to solve.

To solve this problem I use a machine learning based approch.

**Outline :**
1. Data collection and Exploration
2. Data Preprocessing for ML
3. Modelisation and validation 
4. Choice of the king Model
5. Model saving 


# Plugins importing 


In this section, I import all plugin necessary

In [2]:
from flask import Flask, render_template, url_for, request
import pandas as pd
import pickle
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.externals import joblib
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier


/home/jojo/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d


# 1.Data Collection and Exploration

##  1.1 Data Collection 

In this article, I use a YouTube Spam Collection Data Set, it is a public dataset of comments collected for spam research available __[here](https://archive.ics.uci.edu/ml/datasets/YouTube+Spam+Collection)__. It has five datasets composed by 1,956 real messages extracted from five videos that were among the 10 most viewed on the collection period. I will gatherer the fives data set in one. 

In [4]:
#Data collection 
df1 = pd.read_csv("Youtube01-Psy.csv")
df2 = pd.read_csv("Youtube02-KatyPerry.csv")
df3 = pd.read_csv("Youtube03-LMFAO.csv")
df4 = pd.read_csv("Youtube04-Eminem.csv")
df5 = pd.read_csv("Youtube05-Shakira.csv")

#Concatenate all fives data in one
df  = pd.concat([df1, df2,df3,df4,df5])

df


COMMENT_ID                AUTHOR  \
0    LZQPQhLyRh80UYxNuaDWhIGQYNQ96IuCg-AYWqNPjpU             Julius NM   
1    LZQPQhLyRh_C2cTtd9MvFRJedxydaVW-2sNg5Diuo4A           adam riyati   
2    LZQPQhLyRh9MSZYnf8djyk0gEF9BHDPYrrK-qCczIY8      Evgeny Murashkin   
3            z13jhp0bxqncu512g22wvzkasxmvvzjaz04       ElNino Melendez   
4            z13fwbwp1oujthgqj04chlngpvzmtt3r3dw                GsMega   
5    LZQPQhLyRh9-wNRtlZDM90f1k0BrdVdJyN_YsaSwfxc          Jason Haddad   
6            z13lfzdo5vmdi1cm123te5uz2mqig1brz04        ferleck ferles   
7          z122wfnzgt30fhubn04cdn3xfx2mxzngsl40k          Bob Kanowski   
8            z13ttt1jcraqexk2o234ghbgzxymz1zzi04                  Cony   
9            z12avveb4xqiirsix04chxviiljryduwxg0           BeBe Burkey   
10         z13auhww3oufjn1qo04ci3grqqjmfjexxuo0k             Huckyduck   
11             z13xit5agm2zyh4f523rst2gowmbx5bml           Lone Twistt   
12             z13pejoiuozwxtdu323dspopnri4xts0f          Archie Lewis   
13         z121zxaxsq25z5k5o04ch1o5jqqfij3gtm40k       TheUploadaddict   
14         z12oglnpoq3gjh4om04cfdlbgp2uepyytpw0k        Francisco Nora   
15           z13phrmwrkfisn5er22eyrbpbvaiwfvwf04  Gaming and Stuff PRO   
16           z13bgdvyluihfv11i22rgxwhuvabzz1os04           Zielimeek21   
17           z13vxpnoxsyeuv2jr04cctprprb1slnxdf4        OutrightIgnite   
18         z12qth5j0ob1fx3q404chvy4fz32tbkpllk0k        Tony K Frazier   
19           z13etj0bclzfztuwc04cgfvrgmf3fvjor1g         Jose Renteria   
20           z12axnji5w2axxht522thb3bktvqjdlbp04          zhichao wang   
21         z13ozdmr4lf3uzc5z04cix2zkyjzgvcyemw0k       Carlos Thegamer   
22           z12ohdxjtsatvppjb04cctprprb1slnxdf4        OutrightIgnite   
23         z12ntlcqht2bvjewi04cf1up0xjvs5lq3mc0k              Owen Lai   
24   LZQPQhLyRh9EXArr4ZnVcDonSbvSMHKYOT24e_qR6fE           ||GuitarZ||   
25   LZQPQhLyRh9y57URF7qpZRk3MVAJNLNhhZga_5YWBU8         Living4Techno   
26   LZQPQhLyRh9vw01Xvvw5yWzZEUOPG1hSgRMHep55-Yw            8-BitMusic   
27           z13kszcinpnvc34v2234fnpxkpmlw3nhc04            Kyle Jaber   
28         z13tj514otzlurfbc04ccjwhrnmej1iihqw0k         Brandon Pryor   
29         z13zvh1rmk3cf3mby04civbq5mjtddmbysk0k         Fun&amp;Hacks   
..                                           ...                   ...   
340  _2viQ_Qnc6-1oZCLsUWjl3-g4QrWxMQXsSTs2Hy4MGI         damion taylor   
341  _2viQ_Qnc6_YN7xFNAg14zX99Y614Salf57yOcrBRSw       Shadrach Grentz   
342  _2viQ_Qnc6_fyjM2m-ismUToowpNFauwtldKlfjbtIk          Joshua Kasey   
343  _2viQ_Qnc6_AXZ4E5CeA3LHE36RAijd3QKgUI-YvjWI          ricky swaggz   
344  _2viQ_Qnc68ki9xsFeN2y1_ZiHYcZC8Qv1GyHfwqr7Y           steven reed   
345  _2viQ_Qnc69mci30y5muwQXNMaeCmIvZ4ca8l_4zPmA       Johnny Rei Vlog   
346  _2viQ_Qnc68cxHfQzecR1L9f-hbLNIJ7VxpCZtctPMk  yakikukamo FIRELOVER   
347  _2viQ_Qnc6-G71VMp3dR76dfQTcrRHpiNXJh2jm8V_M       Johnny Rei Vlog   
348  _2viQ_Qnc6-3Nk200KmVtS-kiCS_1CjKJsMIbXakyfI           hsn moghrbi   
349  _2viQ_Qnc6-UzHByAP8y3BxG633jelEC_fxtFRUvLSA     Zuzanna Sztandera   
350  _2viQ_Qnc68LpP5gDCaWQuiywObesTUlRgSQExMVMac       Chelsea Fischer   
351          z12xc3ly4x3uttmci22xff24nqqxwb0je04         Lisa Matthews   
352  _2viQ_Qnc6-PfOjDtTwbTalW_5TRtvBKMcHZdDrcI2o         laura elliott   
353  _2viQ_Qnc69GH3FQl348HonbRxpbmtsR5CUei0zkJog         Riley Rollins   
354  _2viQ_Qnc69S12dQyWLf0QBgUD29OMTe71geFOn4PJA         Oona Sarlotta   
355  _2viQ_Qnc69Ic8yEMHsemUQiq01-kwwqnbDowMO9kdM            lynchee360   
356  _2viQ_Qnc6-qHJ_u9Yv84vj4yOAPLUL3ZibCc7b-vBI            FAHAD KHAN   
357  _2viQ_Qnc6_HU65mTzCmXnjA-WLt7XqxqPj7EwAtlO0          ricky swaggz   
358  _2viQ_Qnc6-jk58CPwBnqfbM6oByJH5oPvCtKecLQyo       Shadrach Grentz   
359  _2viQ_Qnc6_fgKR1W7-k1lbVURi8hVbMlQAMSOCSnyk           ThirdDegr3e   
360  _2viQ_Qnc69MEEHHJxZ427KX8MlljJPnUC2YBbvbWwY           ThirdDegr3e   
361  _2viQ_Qnc6_RKHVetk9kLzx8ZC62_J7y73FWFSBTe8Q           ThirdDegr3e   
362  _2viQ_Qnc68TufyXKiTwky80ewSPbhRiD5XFH

## 1.2 Data Exploration

In [5]:
#Dysplay collumn 
list(df.columns)

['COMMENT_ID', 'AUTHOR', 'DATE', 'CONTENT', 'CLASS']

In [6]:
#Plot five fist collumn of data 
df.head()

COMMENT_ID            AUTHOR  \
0  LZQPQhLyRh80UYxNuaDWhIGQYNQ96IuCg-AYWqNPjpU         Julius NM   
1  LZQPQhLyRh_C2cTtd9MvFRJedxydaVW-2sNg5Diuo4A       adam riyati   
2  LZQPQhLyRh9MSZYnf8djyk0gEF9BHDPYrrK-qCczIY8  Evgeny Murashkin   
3          z13jhp0bxqncu512g22wvzkasxmvvzjaz04   ElNino Melendez   
4          z13fwbwp1oujthgqj04chlngpvzmtt3r3dw            GsMega   

                  DATE                                            CONTENT  \
0  2013-11-07T06:20:48  Huh, anyway check out this you[tube] channel: ...   
1  2013-11-07T12:37:15  Hey guys check out my new channel and our firs...   
2  2013-11-08T17:34:21             just for test I have to say murdev.com   
3  2013-11-09T08:28:43   me shaking my sexy ass on my channel enjoy ^_^    
4  2013-11-10T16:05:38            watch?v=vtaRGgvGtWQ   Check this out .   

   CLASS  
0      1  
1      1  
2      1  
3      1  
4      1

In [15]:
#Plot data size
df.shape

(1956, 5)

In [16]:
#Verify null values for collumns
df.isnull().sum()

COMMENT_ID      0
AUTHOR          0
DATE          245
CONTENT         0
CLASS           0
dtype: int64

**Note:** I notice 245 null values for the collumn DATE and Zero for others. The DATE collumn will ne be usefull for the model. So, I will not treat the DATE NULL values cases.

In [11]:
#Features extractions
df_data = df[['CONTENT', 'CLASS']]
df_data.head()

CONTENT  CLASS
0  Huh, anyway check out this you[tube] channel: ...      1
1  Hey guys check out my new channel and our firs...      1
2             just for test I have to say murdev.com      1
3   me shaking my sexy ass on my channel enjoy ^_^       1
4            watch?v=vtaRGgvGtWQ   Check this out .      1

**Note:** I extracted only data for the CONTENT and CLASS collumns because the others collumns are not usefull for the model

In [12]:
#Display the repartition of data in CLASS
counts= df_data["CLASS"].value_counts()
counts

1    1005
0     951
Name: CLASS, dtype: int64

**Note:** I notice that the CLASS is not imbalaced. Thats a good point.

## 2. Data preprocessing for Machine Learning 

In [13]:
# Extract the features 
df_x = df_data['CONTENT']
df_x.head()


0    Huh, anyway check out this you[tube] channel: ...
1    Hey guys check out my new channel and our firs...
2               just for test I have to say murdev.com
3     me shaking my sexy ass on my channel enjoy ^_^ 
4              watch?v=vtaRGgvGtWQ   Check this out .
Name: CONTENT, dtype: object

In [14]:
# Extract the labal
df_y = df_data.CLASS
# Plot five first rows
df_y.head()

0    1
1    1
2    1
3    1
4    1
Name: CLASS, dtype: int64

In [17]:
# Encoding the features with countVectorizer
corpus = df_x
cv = CountVectorizer()
X = cv.fit_transform(corpus)
# Check X size

X.shape
print(X)

  (0, 2335)	1
  (0, 949)	1
  (0, 4037)	1
  (0, 4409)	1
  (0, 3934)	1
  (0, 2949)	1
  (0, 965)	1
  (0, 573)	1
  (0, 2070)	1
  (1, 3769)	1
  (1, 1052)	1
  (1, 2419)	1
  (1, 2394)	1
  (1, 3080)	2
  (1, 3523)	1
  (1, 4284)	1
  (1, 2137)	1
  (1, 2704)	1
  (1, 2705)	1
  (1, 3909)	3
  (1, 4104)	1
  (1, 2208)	1
  (1, 4152)	1
  (1, 1669)	1
  (1, 2947)	1
  :	:
  (1952, 3269)	1
  (1952, 475)	2
  (1952, 810)	1
  (1952, 3666)	1
  (1952, 4052)	1
  (1952, 2499)	1
  (1952, 707)	1
  (1952, 4228)	1
  (1952, 3643)	1
  (1952, 2213)	1
  (1952, 403)	1
  (1952, 1699)	1
  (1952, 2137)	1
  (1952, 2208)	1
  (1952, 3934)	1
  (1953, 4357)	1
  (1954, 4312)	1
  (1954, 3505)	1
  (1954, 595)	1
  (1954, 3623)	1
  (1955, 1205)	1
  (1955, 3505)	1
  (1955, 738)	1
  (1955, 3909)	1
  (1955, 2208)	1


In [19]:
# Split Data into train and test
X_train, X_test, y_train, y_test = train_test_split(X, df_y, test_size = 0.25, random_state = 42)
target_names = ['No Spam', 'Spam']
print(X_train)


  (0, 2137)	1
  (0, 2213)	1
  (0, 1959)	1
  (0, 4277)	1
  (0, 1047)	1
  (0, 735)	1
  (0, 2379)	1
  (1, 3934)	1
  (1, 2762)	1
  (1, 2208)	2
  (1, 2419)	1
  (1, 2281)	1
  (1, 2213)	1
  (1, 3907)	1
  (1, 2469)	1
  (1, 679)	1
  (1, 3266)	1
  (2, 3643)	1
  (2, 2807)	1
  (3, 965)	1
  (3, 2949)	1
  (3, 3934)	1
  (3, 2911)	1
  (3, 4419)	1
  (3, 3072)	1
  :	:
  (1463, 4419)	1
  (1463, 4153)	1
  (1463, 3090)	1
  (1463, 3511)	1
  (1463, 4396)	1
  (1463, 4191)	1
  (1463, 608)	1
  (1463, 3226)	2
  (1463, 1451)	1
  (1463, 4426)	1
  (1464, 655)	1
  (1465, 3909)	1
  (1465, 2419)	1
  (1465, 3966)	1
  (1465, 2848)	1
  (1465, 2533)	1
  (1465, 869)	1
  (1465, 2752)	1
  (1465, 4153)	1
  (1466, 3909)	1
  (1466, 2884)	1
  (1466, 709)	1
  (1466, 3913)	1
  (1466, 2391)	1
  (1466, 3542)	1


# 3. Modelisation and validation 

In this part I will use the following algorithms: Naive Bayes, Logistic Regression, Random Forest and SVM.


**Metrics** 

**Recall:** corresponds to the number of comments correctly attributed to class i compared to the total number of comments belonging to class i (total true positive).

**Precision:** corresponds to the number of comments correctly attributed to class i compared to the total number of comments predicted as belonging to class i (total predicted positive).

**F1-SCore:**  is calculated by harmonic mean Precision and Reccall




##  3.1 Naive Bayes

In [23]:
# Train with Naive Bayes
NaiveBayes= MultinomialNB()
NaiveBayes.fit(X_train, y_train)

# Test Naive Bayes and compute the score


y_predNB =NaiveBayes.predict(X_test)


NaiveBayes.score(X_test, y_test)
# Plot classification report
classification_report(y_test,y_predNB,target_names=target_names)


'             precision    recall  f1-score   support\n\n    No Spam       0.92      0.92      0.92       221\n       Spam       0.94      0.93      0.93       268\n\navg / total       0.93      0.93      0.93       489\n'

##  3.2 Logistic Regression 

In [13]:
# Train with Logistic Regression
logisticRegr = LogisticRegression()
logisticRegr.fit(X_train, y_train)

# Test with Naive Bayes and compute the score
y_predRL =logisticRegr.predict(X_test)
logisticRegr.score(X_test, y_test)

# Plot classification report
classification_report(y_test,y_predRL,target_names=target_names)

'             precision    recall  f1-score   support\n\n    No Spam       0.91      0.98      0.95       221\n       Spam       0.98      0.92      0.95       268\n\navg / total       0.95      0.95      0.95       489\n'

## 3.3 Random Forest

In [14]:
# Train with Random Forest
RandomForest=RandomForestClassifier(n_estimators=100)
RandomForest.fit(X_train,y_train)

# Test with Random Forest and compute the score
y_predRF=RandomForest.predict(X_test)
RandomForest.score(X_test, y_test)

# Plot classification report
classification_report(y_test,y_predRF,target_names=target_names)


'             precision    recall  f1-score   support\n\n    No Spam       0.91      0.99      0.95       221\n       Spam       0.99      0.92      0.95       268\n\navg / total       0.95      0.95      0.95       489\n'

## 3.4 SVM

In [15]:
#Train with SVM
SVM= svm.SVC()
SVM.fit(X_train,y_train)

#Test with SVM and compute the score
y_predSVM=SVM.predict(X_test)
SVM.score(X_test, y_test)

#Plot classification report
classification_report(y_test,y_predSVM,target_names=target_names)

'             precision    recall  f1-score   support\n\n    No Spam       0.50      1.00      0.67       221\n       Spam       1.00      0.18      0.31       268\n\navg / total       0.78      0.55      0.47       489\n'

# 4. Choice of the King Model

My main aim is to detect effectively a spam YOUTUBE comment. That's 1 CLASS value. The best model is Random Forest because it has the higher f1-score(0.96)

# 5. Model Saving

In [16]:
    with open(r"MODEL.pickle", "wb") as output_file:
        pickle.dump(RandomForest, output_file)